### PR TITLE
 -checkaction=context: improve struct equals detection

### PR DESCRIPTION
**Pull request by wilzbach**
Currently with struct comparison only boolean and integer fields are supported as otherwise DMD generates fieldwise comparison and the current context logic doesn't detect this.
This PR improves this. We currently don't have the facilities to test this properly in DMD and as the formatting depends on druntime anyhow, I will add the test to druntime afterwards.

Note: these checks _only_ happen when `-checkaction=context` has been chosen by the user.

-----

Example 1:

```d
struct T2 {
	bool b;
	int i;
	float f1 = 2.5;
	float f2 = 0;
	string s1 = "bar";
	string s2;
}

T2 t1;
T2 t2 = {s1: "bari"};
assert(t1 == t2);
```

->

```
assert failed: T2(b:false, i:0, f1:2.5, f2:0, s1:"bar", s2:"") != T2(b:false, i:0, f1:2.5, f2:0, s1:"bari", s2:"")
```

-----

It also works with negation.
Example 2:

```d
struct T2 {
	bool b;
	int i;
	float f1 = 2.5;
	float f2 = 0;
	string s1 = "bar";
	string s2;
}

T2 t1;
assert(t1 != t1);
```

->

```
assert failed: T2(b:false, i:0, f1:2.5, f2:0, s1:"bar", s2:"") == T2(b:false, i:0, f1:2.5, f2:0, s1:"bar", s2:"")
```

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -5764,6 +5764,161 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+
+        Example:
+            S  { int a; float f;}
+            S s1, s2;
+            s1 == s2;
+            ->
+            s1.a == s2.a && s1.f == s2.f
+
+        Here, we try to conservatively detect such a pattern and return the respective
+        structs as `left` and `right`.
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]: This approach appears to be:

1. compiler converts == to a conjunctive form
2. this code detects a conjunctive form and more or less reverts it back to the original

This seems like a very fragile and backwards approach. It's fragile in that it's highly dependent on exactly what happens with earlier lowerings, and fragile because it can falsely detect patterns written directly by the user.

There's got to be a better way. It's also unclear to me if this PR a problem that even needs a solution.

Comment (wilzbach) [reply src/dmd/expressionsem.d]: >  reverts it back to the original

Well, it doesn't revert it. It just adds the detected structs to `_d_assertfail`, s.t. we can actually get a sensible context message with `-checkaction=context`.

>  It's also unclear to me if this PR a problem that even needs a solution.

`-checkaction=context` is an _existing_ switch that rewrites the assert message, s.t. the user sees the error message directly in the error and doesn't have to rerun the program. Apart from generally improving run/debug ergonomics, this is very important if you have a testsuite that takes a long time or runs on a remote server (e.g. CI or production) where it's hard to reproduce the failure.

tl;dr: with `-checkaction=context` you just get:

```
core.exception.AssertError@foo.d(14): Assertion failure
```

More detailed explanation: https://dlang.org/changelog/2.085.0.html#assert

> This seems like a very fragile and backwards approach. It's fragile in that it's highly dependent on exactly what happens with earlier lowerings,

Yup, but I don't know a better way to do it in DMD.
Also, if these lowerings get ever changed, it would just fallback back to old "assert failed XYZ" where it just stringifies the AST.

> and fragile because it can falsely detect patterns written directly by the user.

No, it won't have any false positives with this approach as it only adds the structs to the assert message if every part of the expression consists of binary expressions of both of them. 
A few examples of how DMD behaves _currently_ with `-checkaction=context`:

1) Field comparison

```
assert(t1.a == t2.a);
->
core.exception.AssertError@foo.d(14): 0 != 2
```

2) Struct comparison

```
assert(t1 == t2);
->
assert(t1 == t2) failed
```

3) User-generated struct comparison

```
assert(t1.a == t2.a && t1.f == t2.f);
->
assert(t1.a == t2.a && (t1.f == t2.f)) failed
```

Now with this PR (2) would result in e.g.:

```diff
-assert(t1 == t2) failed
+T2(a:0, f:nan, s:"") != T2(a:2, f:nan, s:"")
```

Note that the code currently doesn't even rewrite (3), though if it did that would even be an improvement.

> There's got to be a better way

I'm all open for better solutions.

**Comment by WalterBright** [comment]
> I'm all open for better solutions.

Let's step back a bit. The expression `assert(a == b)` generates a message from the template `_d_assert_fail!"=="(a, b)`. Rather than building all the message generating logic into the compiler, this looks like an ideal application for "expression templates" which can then be pushed into druntime.

https://dlang.org/articles/templates-revisited.html

**Comment by wilzbach** [comment]
> The expression assert(a == b) generates a message from the template _d_assert_fail!"=="(a, b). Rather than building all the message generating logic into the compiler, this looks like an ideal application for "expression templates" which can then be pushed into druntime.

Huh? This is already in druntime:

https://github.com/dlang/druntime/blob/master/src/object.d#L4326

The problem is that by the time the compiler looks at the `assert` expression, it has already done a lot of other lowering.

**Comment by WalterBright** [comment]
I know it's there, I was just thinking about extending it. It only takes one operator.

As for lowering, seriously, does it matter? Why can't _d_assert_fail be overloaded so it can accept a sequence of && expressions? I haven't thought this through, just looking for a more generic scheme.

**Comment by Geod24** [comment]
No tests ?
I'd really like this feature to work, as I believe it's a MAJOR leap in D usability, but it does seem that the code will be brittle.

**Comment by wilzbach** [comment]
We don't have tests here as this would lock DMD and druntime even more, but we have them e.g. here:

https://github.com/dlang/druntime/pull/2720

**Comment by wilzbach** [comment]
> No tests ?

Added a potential test PR: https://github.com/dlang/druntime/pull/2726

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -5796,22 +5956,40 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+
+                        // runtime args
+                        (*es)[0] = structEqualsLeft;
+                        (*es)[1] = structEqualsRight;
+                    }
                     // structs with opEquals get rewritten to a DotVarExp:
                     // a.opEquals(b)
                     // https://issues.dlang.org/show_bug.cgi?id=20100
-                    if (args.length == 1)
-                    {
```
Comment (Geod24) [inline src/dmd/expressionsem.d]: Accidental removal ?

**Comment by Geod24** [comment]
@MoonlightSentinel : What should we do with this ?

**Comment by MoonlightSentinel** [comment]
The mentioned tests cases work already, so I think we can close this.
